### PR TITLE
Preserving openHook callbacks on SchemeSpouts

### LINKED ISSUE #68: SchemeSpout doesn't handle callOnOpen properly
This manifests in three ways.

If we do simply:

``` scala
val hookedSpout = schemeSpout.openHook { _ => println("openHook") }
```

We are given back a `BaseSpout` that is not a `SchemeSpout`, and we are back to the `poll`-based spout implementation (which, incidentally, will always be `Nil`).

This can be worked around, sort of, by using a `FlatMappedSchemeSpout`, because that implements `openHook`.  But there are problems there too.

If we have code that looks like:

``` scala
val hookedSpout = fmSchemeSpout.openHook { _ => println("openHook") }
val mappedSpout = hookedSpout.flatMap { x => Some(x) }
```

`mappedSpout` will not honor the hook, because there is no default implementation in `FlatMappedSchemeSpout` for `callOnOpen`.

Furthermore, code like this:

``` scala
val hookedSpout = fmSchemeSpout.openHook { _ => println("openHook") }
val mappedSpout = hookedSpout.flatMap { x => Some(x) }
val hookedSpout2 = mappedSpout.openHook { _ => println("openHook2") }
val hookedSpout3 = hookedSpout2.openHook { _ => println("openHook3") }
```

`hookedSpout3` will now honor the first hook, and the third, but not the second.

### PR DESCRIPTION
**Pull request by jnievelt**
Fixes #68 (simpler approach)

### CODE REVIEW

File: tormenta-core/src/test/scala/com/twitter/tormenta/spout/SchemeSpoutTest.scala [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+      val richSpout = hookedSpout.getSpout
+      spout.fn(null)
+
+      assert(hookCalled)
+      assert(spout.transformed == Seq(1, 2, 3))
+    }
+
+    "calls multiple open hooks" in {
+      var hook1 = false
```
Comment (johnynek) [inline tormenta-core/src/test/scala/com/twitter/tormenta/spout/SchemeSpoutTest.scala]: do you want to use `var hook1 = 0` and then `hook1 = hook1 + 1` and make sure we wind up with `1`? This isn't checking for multiple calls.

**Comment by johnynek** [comment]
looks good, just a minor comment about maybe being more strict in the test.

File: tormenta-core/src/test/scala/com/twitter/tormenta/spout/SchemeSpoutTest.scala (hunk 1 above)
Comment (jnievelt) [inline tormenta-core/src/test/scala/com/twitter/tormenta/spout/SchemeSpoutTest.scala]: The closures that set these also check for multiple calls.  But I can switch it out for `AtomicInteger`, etc., if that's cleaner.

**Comment by johnynek** [comment]
indeed, sorry I missed that!

👍

**Pull request merged**